hysds/utils.py

```diff
@@ -737,3 +737,143 @@ def mark_localized_datasets(job, ctx):
+    with open(file_path, 'r') as f:
+        checksum = f.readline().rstrip('\n')  # checksum file is only 1 line, for some reason it adds \n at the end
+        return checksum
+
+
+def generate_list_checksum_files(job):
+    '''
+    :param job:
+    :param cxt:
+    :return: list of all checksum files, so we can compare one by one
+             ex. list of dictionaries: [ {'file_path': '/home/ops/hysds/...', 'algo': 'md5'}, { ... } ]
+    '''
+    # reusing directory code from the localize_urls() function
+    job_dir = job['job_info']['job_dir']  # get job info
+
+    files_with_checksum = []
+    for i in job['localize_urls']:
+        url = i['url']
+        path = i.get('local_path', None)
+        cache = i.get('cache', True)
+        if path is None:
+            path = '%s/' % job_dir
+        else:
+            if path.startswith('/'):
+                pass
+            else:
+                path = os.path.join(job_dir, path)
+        if os.path.isdir(path) or path.endswith('/'):
+            path = os.path.join(path, os.path.basename(url))
+        dir_path = os.path.dirname(path)
+
+        if os.path.isdir(path):  # if path is a directory, loop through each file in directory
+            for file in os.listdir(path):
+                full_file_path = os.path.join(path, file)
+                hash_algo = check_file_is_checksum(full_file_path)
+                if hash_algo:
+                    files_with_checksum.append({'file_path': file, 'algo': hash_algo})
+        else:  # if path is a actually a file
+            hash_algo = check_file_is_checksum(path)
+            if hash_algo:
+                files_with_checksum.append({'file_path': path, 'algo': hash_algo})
+    return files_with_checksum
+
+
+def validate_checksum_files(job, cxt):
+    '''
+    :param job: _job.json
+    :param cxt: _context.json
+    :return: void, will raise exception if localized files have mismatched checksum values
+    '''
+    # list of dictionaries: ex. [ {'file_path': '/home/ops/hysds/...', 'algo': 'md5'}, { ... } ]
+    files_to_validate = generate_list_checksum_files(job)
+
+    mismatched_checksums = []
+    exception_string = 'Files with mismatched checksum:\n'
+
+    for file_info in files_to_validate:
+        algo = file_info['algo']
+        file_path_checksum = file_info['file_path']
+        # this has the hash extension to the file, we need to remove it
+        file_path = file_path_checksum.replace('.' + algo, '')
+
+        calculated_checksum = calculate_checksum_from_localized_file(file_path, algo)
+        pre_computed_checksum = read_checksum_file(file_path_checksum)
+
+        if calculated_checksum.lower() != pre_computed_checksum.lower():
+            mismatched_checksums.append(file_path)
+            exception_string += ('%s: calculated checksum: %s, pre-computed checksum: %s\n' % (file_path, calculated_checksum, pre_computed_checksum))
+
+    if len(mismatched_checksums) > 0:
+        raise Exception(exception_string)
```

Comment:
Return an explicit `True` if no mismatched checksums were found.

Reply:
Yes, i tested the functions individually in a local Docker container, also added unit test file `test/test_preprocessing_checksum.py`